requests-for-comments/2024-01-23-project-environments.md

```diff
@@ -0,0 +1,189 @@
+
+To support the most common workflow of drafting dashboards/etc. in dev/staging and then "promoting" to production, entities such as dashboards would be environment-specific by default, but could be rolled out to the whole project with two clicks. When rolled out, the entity's dependencies would be made project-wide as well (e.g. for a cohort, all actions and other cohorts used in the definition). This would involve no copying though, so there would be no effort needed to sync things.
+
+To support large customers with access control needs, we'd need to add environment-specific project membership to the existing project access control system.
+
+To support the common feature flags workflow of rollout conditions being different between dev and production, we'd make flag rollout groups always environment-specific.
+
+To ensure straightforward yet complete separation of data at the ingestion level (incl. persons), *as well as* steady querying performance, each environment would have its own `team_id`. Due to this requirement, environments would have to be predefined by users and not created at ingestion time, thus each environment would also have its own SDK key (aka "project API key" currently). That seems like a safe choice in terms of the UX.
+
+By default, a project would only have one environment implicitly named "Production". A "New environment" button would allow adding a new one, with the environments feature explained in the modal. The current environment would then be selectable from the project selector, roughly looking like this (courtesy of Cory):
+
+<img width="939" alt="image" src="https://github.com/PostHog/meta/assets/4550621/ec966c6b-0a1c-4cb0-a593-0234f4807ab1">
+
+### Downsides
+
+Of course, nothing comes for free: adding evironments has its own downsides. Here, it's the amount of effort required – this feature would touch every part of the app, plus ingestion. It would realistically be a few person-sprints of work (as always, it's hard to estimate accurately).
+
+### Pricing
+
+Environments should start out as a paid feature, since they're key for professional users, and more of a nice-to-have for hobbyists. I would suggest it falling under the same restrictions as multiple projects, i.e. multiple environments being available after putting in a card.
+
+### Rough scope of work
+
+1. Backend: We add an `Environment` model (`posthog_environment` table, fields `id`, `team`, `data_team_id`, `name`, `api_token`) and hook it up to a new project-scoped viewset (`/api/projects/:id/environments/`). Every time a new environment is created, we increment the counter of the `Team.id` field and use the obtained value as the environment's `data_team_id`.
+1. Backend: For every existing project (`Team`) we create a matching default `Environment`, reusing the project's `id` (for `data_team_id`) and `api_token` (for `api_token`). Any time a new project is created, it also gets a default environment based on the same logic.
```

Comment:
For maximal flexibility we could instead do this (note that we should hash out the requirements more fully before to better understand the needs):

for each sharable item (feature_flag, insight, ...) we'll change them to instead of having a `team_id` to have an `organization_id` and `sharing_key` (that's a UUID unique across everything <- though technically unique across organizatisation is good enough)

Then we'll have a new table `shared_across_projects`, that has 
1. organization_id
2. team_id
3. sharing_key

We could then have configurations like this:

Imagine the `shared_across_projects` table looks like this:
| organization | team | sharing_key |
|-|-|-|
| PostHog | Website-EU | 1234 |
| PostHog | Website-US | 2233 |
| PostHog | Website-EU | 5555 |
| PostHog | Website-US | 5555 |

we can have release annotations being separate, because they hit production at different times, but some annotations are shared
| Annotation name | organization | sharing_key | timestamp
|-|-|-|-
| release-380 | PostHog | 1234 | 15:00
| release-380 | PostHog | 2233 | 15:10
| Christmas holidays | PostHog | 5555 | dec 24th

We can have dashboard & insights that are shared together across projects (note I'm re-using the sharing key from above, but it could be a different sharing key)
| dash name | organization | sharing_key | ...
|-|-|-|-
| Overview | PostHog | 5555 | 
| EU only dash | PostHog | 1234 |
| US only dash | PostHog | 2233 |

| insight name | organization | sharing_key | ...
|-|-|-|-
| Overview first insight | PostHog | 5555 | 
| Overview second insight | PostHog | 5555 | 
| EU only dash first insight  | PostHog | 1234 |
| EU only dash second insight  | PostHog | 1234 |
| US only dash first insight  | PostHog | 2233 |

Then when looking up dashboards for the project - we'll query all the dashboards for the organization & all the sharing keys for the organization and join over them in the `shared_across_projects`. If this is too slow we can denormalise some info somewhere as a quicker read path. Optionally we can add a product field (dashboards, feature flags, ...) to the `shared_across_projects` for speeding up lookups.

In the UI the user could say 
1. share everything across everything - in which case we'd just use a single sharing key everywhere (note that we could do splits later, by overwriting the sharing keys in the tables)
2. all my dashboards should be share across all my projects, but everything else should be separate - cool, we just use the single sharing key for all dashboards & unique by project sharing keys for everything else. 
3. The user from above can later pick that I want to create a dashboard that isn't shared (by creating a new sharing key in the backend). They can then later choose that this dashboard should be shared with some other projects (we can add entries to the `shared_across_projects` table to make that happen ... note that we'd want to think through how it works for related objects, i.e. insights on a dashboard).
4. during ingestion we'd look up by team_id, sharing_key for group_type creation - we can also denormalise this (duplicate the data), for it to be faster to access

Reply:
What you're outlining is exactly what proposal B "Dashboards/insights/etc. sharing between projects" would be. However, as the RFC points out, that runs into the problem of taxonomy being different projects, and especially group types.
I'm not sure what you mean by "during ingestion we'd look up by team_id, sharing_key for group_type creation" and if that solves the problem. Can you clarify?

This is definitely not feasible: "I'd start with seeing if this model works for dashboards and insights & then slowly add it to more products" – at the very least the set of shareable entity types has to be: dashboards, insights, cohorts, actions. This also means you couldn't say "all my dashboards should be share across all my projects, but everything else should be separate" – no, if dashboard X is shared across projects, then all its dependencies must now be shared across those same projects. And the important complication is that it can't be "use only in this project" or "share across all projects", because then if you want to run two  _completely separate_ projects in one org, you can no longer do that. But let me know if I misunderstood what you meant on this point.